airflow/jobs/scheduler_job.py

```diff
@@ -797,15 +797,26 @@ def process_file(self, file_path, zombies, pickle_dags=False, session=None):
         # process and due to some unusual behavior. (empty() incorrectly
         # returns true as described in https://bugs.python.org/issue23582 )
         ti_keys_to_schedule = []
+        refreshed_tis = []
 
         self._process_dags(dagbag, dags, ti_keys_to_schedule)
 
-        for ti_key in ti_keys_to_schedule:
-            dag = dagbag.dags[ti_key[0]]
-            task = dag.get_task(ti_key[1])
-            ti = models.TaskInstance(task, ti_key[2])
+        # Refresh all task instances that will be scheduled
+        TI = models.TaskInstance
+        filter_for_tis = ([and_(TI.dag_id == dag_id,
+                                TI.task_id == task_id,
+                                TI.execution_date == execution_date)
+                           for dag_id, task_id, execution_date, _ in ti_keys_to_schedule])
+
+        if filter_for_tis:
+            refreshed_tis = session.query(TI).filter(or_(*filter_for_tis)).with_for_update().all()
```

Comment:
This clause is interesting because if the `filter_for_tis` then `or_` is a tautology so all TIs are returned. It's worth to check if it occurs in other places.